### DISCUSSION TITLE
Comparing Time Series Model Performance | Basics

### DISCUSSION THREAD

**Discussion by ngupta23** [pycaret/pycaret#1808] [FAQs]
# Background

When starting any modeling task, it is often advisable to build a baseline model to benchmark subsequent models against. This baseline model could be a naive model that takes the mean of all data points. It could be a model that just forecasts the last known data point. Or it could be any model that incorporate the current state of the art that you wish to beat. 

```python
# Seasonal Naive Baseline ----
baseline = exp.create_model("snaive")
exp.plot_model(baseline)
```

![image](https://user-images.githubusercontent.com/33585645/140608976-eda3ee07-0533-4a4c-a596-bfd0e0977a1b.png)

Following this, more intelligent models can be built with the hope of improving the performance over the baseline model(s). This process is usually iterative e.g. you may want to compare multiple modeling strategies against the baseline.

It is also important to note that for a fair comparison, we should

1. Make sure we are performing cross-validation so that the best models selection is not influenced by just a small time frame where an otherwise good model might perform poorly or vice versa. In general, cross-validation leads to a selection of the best model that is more likely to generalize in the future.
2. When performing cross-validation across model, we want to make sure that we do a fair comparison. Hence, the folds used to perform cross validation must be consistent between models. 

Luckily, `pycaret` automatically takes care of this by default. As you can see, the folds use the same train/test `cutoff` across the 3 folds in this setup.

```python
#### Classical Statistical Model: ARIMA
model1 = exp.create_model("arima")
exp.plot_model(model1)

#### Classical Statistical Model: ETS
model2 = exp.create_model("ets")
exp.plot_model(model2)

#### Regression Based Model with Lags: lightgbm
model3 = exp.create_model("lightgbm_cds_dt")
exp.plot_model(model3)
```
![image](https://user-images.githubusercontent.com/33585645/140609402-b71ad942-3119-47f4-885e-6fddc067becf.png)
![image](https://user-images.githubusercontent.com/33585645/140609424-caeae655-1587-4582-aa89-127fe488bf52.png)
![image](https://user-images.githubusercontent.com/33585645/140609437-85c07b0d-eb55-42f8-8520-9afdefe062e0.png)

So some models do better than others. How do I know which one forecasts the best? While certainly possible to do this manually, it can take a while to accomplish. That's why `pycaret` provides a convenient wrapper to do this - `compare_models`. At the same time, it provides user customization so that experienced users can adapt this to their needs. Let's see how it works.

NOTE: A complete notebook example for comparing models can be **[found here](https://gist.github.com/ngupta23/f94b2f3ce22b7a0203446b77f2972ffa)**.

# Automated Comparison

```python
best_baseline = exp.compare_models()
exp.plot_model(best_baseline)
```
![image](https://user-images.githubusercontent.com/33585645/140609498-c0436e51-26ea-4f7f-8ab5-4c4aa98279f1.png)
![image](https://user-images.githubusercontent.com/33585645/140609508-bb41d65a-572b-4553-a008-721323b0c1f5.png)

The grid that you see is essentially showing the mean results across the cross-validation folds for the various models. If you look carefully, the results for `arima`, `ets` and `lightgbm_cds_dt` should match the mean results from the manual flow above.

Also, by default the `compare_models` call will return the best model back from the list so it can be further tuned to improve model performance. You can read about tuning [here](https://github.com/pycaret/pycaret/discussions/1791) and [here](https://github.com/pycaret/pycaret/discussions/1795).

Also, in some cases, you may wish to ensemble different models so you may want to get the top N models back. This can be accomplished by using the `n_select` argument.

```python
#### Return the 3 top models ----
best_baseline_models = exp.compare_models(n_select=3)
for model in best_baseline_models:
  exp.plot_model(model)
```

![image](https://user-images.githubusercontent.com/33585645/140609966-688e001f-26a9-4f18-a7da-5c28d6fad477.png)
![image](https://user-images.githubusercontent.com/33585645/140609977-50dc9df1-ffe0-42f4-af2f-33e80c05790b.png)
![image](https://user-images.githubusercontent.com/33585645/140609989-c8d32bab-12fd-4ed1-bee8-20f02a9ea88a.png)

That is how simple it is to compare the performance of multiple models with `pycaret`. That is it for this FAQ. Be sure to read the suggested next reads to see how you can customize the comparison process to suit your specific needs.

Happy Forecasting!

# Suggested Next Reads

- https://github.com/pycaret/pycaret/discussions/1809
- https://github.com/pycaret/pycaret/discussions/1810
- https://github.com/pycaret/pycaret/discussions/1761

**Comment by moezali1** [comment]
PyCaret 4.0 sunset the legacy time-series module (sktime/prophet integration), so this FAQ no longer maps to a supported code path.

If you still need time-series forecasting, use sktime / prophet / statsforecast directly — those projects have surpassed what the PyCaret wrapper offered.

For 4.0 docs, see https://pycaret.org/docs and the cleanup notice at https://github.com/pycaret/pycaret/issues/4177.

Closing as outdated.
